### PR TITLE
NFS client provisioner

### PR DESCRIPTION
**Pull request by verwilst**
**What this PR does / why we need it**:
Adds nfs-client-provisioner chart
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by verwilst** [comment]
I've signed the CLA in the meantime.

**Comment by verwilst** [comment]
Should obsolete #2779

**Comment by verwilst** [comment]
I've based some parts of the README from PR above though.

**Comment by verwilst** [comment]
I've mentioned Kubernetes 1.7+, but since I'm using quite recent api versions, maybe 1.9+ is more suited?

**Review by mlaccetti — COMMENTED** [review]
A few things:

1) `incubator` has been somewhat deprecated, `stable` is a good place to start
2) kubernetes 1.9+
3) provisioner name probably should be empty by default, since afaik there is no "standard" for this

**Comment by verwilst** [comment]
Updated point 1 and 2. About 3. I feel like a 'default' is needed, so I took a look at nfs-server-provisioner. They do it as follows:

```
{{- define "nfs-provisioner.provisionerName" -}}
{{- if .Values.storageClass.provisionerName -}}
{{- printf .Values.storageClass.provisionerName -}}
{{- else -}}
cluster.local/{{ template "nfs-provisioner.fullname" . -}}
{{- end -}}
{{- end -}}
```
But i do not like the 'cluster.local/' prefix. I don't have a single cluster named 'cluster.local' for example so we shouldn't assume that's the name. Afaik there is no way to get the cluster name through Helm.
Any other ideas?

**Comment by mlaccetti** [comment]
`cluster.local` is the default DNS for k8s, so it isn't a particularly bad default.

**Review by mlaccetti — APPROVED** [review]
lgtm

**Comment by kingdonb** [comment]
What are the checks that are holding this patch up? Not sure how to read e2e CI and tide results

**Comment by mlaccetti** [comment]
@kingdonb Waiting on an approval - @unguiculus - can you take a quick peek?

File: stable/nfs-client-provisioner/templates/deployment.yaml
Comment (unguiculus) [inline stable/nfs-client-provisioner/templates/deployment.yaml]: `apps/v1`

**Review by unguiculus — CHANGES_REQUESTED** [review]
A few more things:
* Change `incubator` references to `stable`
* Add `heritage` labels
* Apply our RBAC best practices: https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

**Comment by verwilst** [comment]
@unguiculus Could you review my changes please?

**Comment by wstrange** [comment]
bump

**Comment by wstrange** [comment]
A couple of issues:

clusterrolebinding.yaml uses `name: {{ template "nfs-client-provisioner.name" . }}` for the subject service account name, while deployment.yaml uses:

 `serviceAccountName: {{ if .Values.rbac.create }}{{ template "nfs-client-provisioner.fullname" . }}{{ else }}{{ .Values.rbac.serviceAccountName | quote }}{{ end }}`

They should both use fullname or name - otherwise the service account does not match the cluster role binding.

Also - the  README specifies `.Values.nfs.provisionerName`, but the actual value is 
`.Values.storageClass.provisionerName`

**Comment by verwilst** [comment]
All fixed. Let's get this approved and merged please..

File: stable/nfs-client-provisioner/templates/deployment.yaml
Comment (unguiculus) [inline stable/nfs-client-provisioner/templates/deployment.yaml]: `{{ template "nfs-client-provisioner.serviceAccountName" }}`

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by unguiculus** [comment]
@verwilst Could you check the CI logs? For some reason it did not fail but there's an issue with the deployment in the log.

**Comment by verwilst** [comment]
pull-charts-e2e succeeds now!

File: stable/nfs-client-provisioner/templates/clusterrolebinding.yaml
Comment (davidkarlsen) [inline stable/nfs-client-provisioner/templates/clusterrolebinding.yaml]: shouldn't this use the template for service account name too?

**Comment by verwilst** [comment]
Good catch, fixed.

**Comment by verwilst** [comment]
Anything else that blocks this from being included?

**Comment by davidkarlsen** [comment]
says "W0813 20:42:34.465] Error: release nfs-client-provisioner-y37bwb8jkrwqzvcl-0 failed: storageclasses.storage.k8s.io "nfs-client" already exists
"

**Comment by verwilst** [comment]
Any idea how to work around this? The storageclass shouldn't already exist?

**Comment by davidkarlsen** [comment]
@unguiculus @sameersbn @prydonius can you help us on this one? Storageclass is not a namespaced resource - can there be leftovers between tests?

**Comment by davidkarlsen** [comment]
now it does not fail on existing class, but due to the vol not being mountable/creatable: 
"I0814 04:37:05.423]   Warning  FailedMount  1m  kubelet, gke-jenkins-default-pool-51cafeba-718t  MountVolume.SetUp failed for volume "nfs-client-root" : mount failed: exit status 1" which is no wonder since that would require NFS to run on the test-bench

File: stable/nfs-client-provisioner/templates/clusterrole.yaml
Comment (davidkarlsen) [inline stable/nfs-client-provisioner/templates/clusterrole.yaml:11]: should endpoints access be added too? see: https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/deploy/auth/clusterrole.yaml

Comment (verwilst) [reply stable/nfs-client-provisioner/templates/clusterrole.yaml:11]: Only been added 2 weeks ago so it seems https://github.com/kubernetes-incubator/external-storage/commit/d46083d75be3c046c5313c03c0b4b5f29d9b2ec2. I can sync it though. ;-) Man I really want to get this merged. :-p

Comment (davidkarlsen) [reply stable/nfs-client-provisioner/templates/clusterrole.yaml:11]: That would be good - it was malfunctioning for me without it

**Comment by davidkarlsen** [comment]
/lgtm
/assign @mlaccetti 
/assign @foxish 
/assign @lachie83 

@mlaccetti 
@foxish 
Can we get this one through please?

File: stable/nfs-client-provisioner/README.md
Comment (davidkarlsen) [inline stable/nfs-client-provisioner/README.md]: now this got out of sync

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

**Comment by davidkarlsen** [comment]
@unguiculus @prydonius 
Any chance of getting this one in? It would be very nice to have it.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**
